dask/dataframe/_pyarrow.py

```diff
@@ -76,5 +72,18 @@ def to_pyarrow_string(df):
     if (is_dataframe_like(df) or is_series_like(df)) and is_object_string_index(
         df.index
     ):
-        df.index = df.index.astype(pd.StringDtype("pyarrow"))
+        if isinstance(df.index, pd.MultiIndex):
+            levels_to_set = []
+            level_indices = []
+            for i, level in enumerate(df.index.levels):
+                if is_object_string_dtype(level.dtype):
+                    levels_to_set.append(level.astype(pd.StringDtype("pyarrow")))
+                    level_indices.append(i)
+            if levels_to_set:
+                # set verify_integrity=False to preserve index codes
+                df.index = df.index.set_levels(
+                    levels_to_set, level=level_indices, verify_integrity=False
+                )
```

Comment:
This seems to be a bit more simple

```diff
diff --git a/dask/dataframe/_pyarrow.py b/dask/dataframe/_pyarrow.py
index 51eff76fc..3b7f390cf 100644
--- a/dask/dataframe/_pyarrow.py
+++ b/dask/dataframe/_pyarrow.py
@@ -73,17 +73,15 @@ def to_pyarrow_string(df):
         df.index
     ):
         if isinstance(df.index, pd.MultiIndex):
-            levels_to_set = []
-            level_indices = []
-            for i, level in enumerate(df.index.levels):
-                if is_object_string_dtype(level.dtype):
-                    levels_to_set.append(level.astype(pd.StringDtype("pyarrow")))
-                    level_indices.append(i)
-            if levels_to_set:
-                # set verify_integrity=False to preserve index codes
-                df.index = df.index.set_levels(
-                    levels_to_set, level=level_indices, verify_integrity=False
-                )
+            levels = {
+                i: level.astype(pd.StringDtype("pyarrow"))
+                for i, level in enumerate(df.index.levels)
+                if is_object_string_dtype(level.dtype)
+            }
+            # set verify_integrity=False to preserve index codes
+            df.index = df.index.set_levels(
+                levels.values(), level=levels.keys(), verify_integrity=False
+            )
         else:
             df.index = df.index.astype(pd.StringDtype("pyarrow"))
     return df
```

Also, I don't think we need the `if levels_to_set:` check since we're in a a block where `is_object_string_index(df.index)` is `True`

Reply:
It's simpler, but `dict.keys()` and `dict.values()` are not guaranteed to be sorted.

Reply:
..actually they are since python 3.7. Well then!